go/vt/tabletserver/splitquery/split_params.go

```diff
@@ -0,0 +1,80 @@
+package splitquery
+
+import (
+	"fmt"
+
+	"github.com/youtube/vitess/go/vt/schema"
+	"github.com/youtube/vitess/go/vt/sqlparser"
+)
+
+// SplitParams stores the parameters for splitting a query.
+type SplitParams struct {
+	sql                 string
+	bindVariables       map[string]interface{}
+	splitColumns        []string
+	splitCount          int64
+	numRowsPerQueryPart int64
+
+	// Will contain the AST for the SELECT query.
+	selectAST *sqlparser.Select
+}
+
+// NewSplitParams returns a new SplitParams object.
+// Parameters:
+// 'sql' is the SQL query to split. The query must satisfy the restrictions found in the
+// documentation of the vtgate.SplitQueryRequest.query protocol buffer field.
+// 'bindVariables' are the bind-variables for the sql query.
+// 'splitColumns' the list of splitColumns to use. These must adhere to the restrictions found in
+// the documentation of the vtgate.SplitQueryRequest.split_column protocol buffer field.
+// If splitColumns is nil, the split columns used are the primary key columns (in order).
+func NewSplitParams(
+	sql string,
+	bindVariables map[string]interface{},
+	splitColumns []string,
+	schema map[string]*schema.Table) (*SplitParams, error) {
+	statement, err := sqlparser.Parse(sql)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"splitter.NewSplitParams(): failed parsing query: '%v', err: '%v'", sql, err)
+	}
+	selectAST, ok := statement.(*sqlparser.Select)
+	if !ok {
+		return nil, fmt.Errorf("splitquery.NewSplitParams(): not a select statement")
+	}
+	if selectAST.Distinct != "" || selectAST.GroupBy != nil ||
+		selectAST.Having != nil || len(selectAST.From) != 1 ||
+		selectAST.OrderBy != nil || selectAST.Limit != nil ||
+		selectAST.Lock != "" {
+		return nil, fmt.Errorf("splitquery.NewSplitParams(): unsupported query")
+	}
+	if len(selectAST.From) != 1 {
+		return nil, fmt.Errorf("splitquery.NewSplitParams(): query must have exactly one table"+
+			" in FROM clause. Got: %d", len(selectAST.From))
+	}
```

Comment:
Hmm! Not so much?!
On Feb 27, 2016 12:27 AM, "dumbunny" <notifications@github.com> wrote:

> In go/vt/tabletserver/splitquery/split_params.go
> <https://github.com/youtube/vitess/pull/1526#discussion_r54328823>:
>
> > +			"splitter.NewSplitParams(): failed parsing query: '%v', err: '%v'", sql, err)
> > +	}
> > +	selectAST, ok := statement.(*sqlparser.Select)
> > +	if !ok {
> > +		return nil, fmt.Errorf("splitquery.NewSplitParams(): not a select statement")
> > +	}
> > +	if selectAST.Distinct != "" || selectAST.GroupBy != nil ||
> > +		selectAST.Having != nil || len(selectAST.From) != 1 ||
> > +		selectAST.OrderBy != nil || selectAST.Limit != nil ||
> > +		selectAST.Lock != "" {
> > +		return nil, fmt.Errorf("splitquery.NewSplitParams(): unsupported query")
> > +	}
> > +	if len(selectAST.From) != 1 {
> > +		return nil, fmt.Errorf("splitquery.NewSplitParams(): query must have exactly one table"+
> > +			" in FROM clause. Got: %d", len(selectAST.From))
> > +	}
>
> You are already checking len(selectAST.From) in the previous if statement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/youtube/vitess/pull/1526/files#r54328823>.
>